---
title: "A brief interrogation of circadian rhthym-related genes in Depmap melanoma data"
author: "Theo Killian"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r, echo = FALSE, results = 'asis'}
suppressPackageStartupMessages(library("dplyr"))
knitr::opts_chunk$set(collapse=TRUE, comment="#>", warning=FALSE, message=FALSE)
```

```{r libraries, echo=FALSE, message=FALSE}
## TODO also COSMIC and mutation types, multiomic stratification

library("dplyr")
library("tidyr")
library("tibble")
library("ggplot2")
library("viridis")
library("stringr")
library("ggrepel")
library("ggpubr")
library("depmap")
library("ExperimentHub")
library("biomaRt")
library("plotly")
library("pheatmap")
library("DT")
# library("ggbio")
# library("gridExtra")
# library("ggbeeswarm")
# library("ggVennDiagram")
# library("GenomicRanges")
# https://bioconductor.org/packages/devel/data/experiment/vignettes/depmap/inst/doc/using_depmap.html
my_cols <- c("#0D0887FF", "#6A00A8FF", "#B12A90FF", "#E16462FF", "#FCA636FF", "#F0F921FF")
```

# Introduction

This HTML report details an exploration of the nature of circadian genes in 
melanoma cell lines from the Broad Institute [Depmap](https://depmap.org/portal/data_page/?tab=allData)
datasets using the [depmap R package](https://www.bioconductor.org/packages/release/data/experiment/html/depmap.html),
cancer dependency data described by [Tsherniak, Aviad, et al. "Defining a cancer dependency map." Cell 170.3 (2017): 564-576.](https://www.ncbi.nlm.nih.gov/pubmed/28753430). The DepMap data enables
us to look at CRISPR essentiality, copy number, mutation calls, and transcript
expression for a large panel of cancer cell lines, including melanoma.

## Gene lists

We will interrogate Depmap data using a curated list of genes that regulate the 
human circadian rhythm, which were taken from the publication [The Circadian Clock Component RORA Increases Immunosurveillance in Melanoma by Inhibiting PD-L1 Expression, Liu, et al. 2024](https://pmc.ncbi.nlm.nih.gov/articles/PMC11247325/).

```{r gene_lists}
circadian_gene_list <- c(
  "RORA", "PER3", "PER1", "CRY2", "RORB", "PER2", "NR1D2", "RORC", "CRY1",
  "ARNTL", "NPAS2", "CLOCK", "NR1D1", "ARNTL2")

# driver_gene_list <- c(
#   "BRAF", "NRAS", "NF1", "CDKN2A", "PTEN", "TP53", "TERT", "KIT", "GNAQ",
#   "GNA11", "MITF", "MC1R", "CDK4")
```

### Biomart

Here, we connect to Ensembl via biomaRt (disabled by `eval=FALSE` so it doesn't
automatically run each time). This code retrieves gene metadata (e.g., gene
names, coordinates) and saves it as an `.rds` file.

```{r eval=FALSE}
## Connect to the Ensembl database
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

## get information for all human genes (to be used later)
getBM(
  attributes = c(
  "external_gene_name", "description", "entrezgene_id", "ensembl_gene_id",
  "chromosome_name", "start_position", "end_position"),
      mart = ensembl) %>%
  dplyr::rename(gene_name = external_gene_name,
                entrez_id = entrezgene_id,
                ensembl_id = ensembl_gene_id,
                chromosome = chromosome_name,
                start = start_position,
                end = end_position) %>%
  dplyr::filter(stringr::str_length(gene_name) > 1,
                stringr::str_length(chromosome) < 3) %>%
  dplyr::mutate(description = gsub("\\[Source:.*", "", description)) %>%
  as.data.frame() -> human_genes

saveRDS(human_genes, file = "./data/human_genes.rds")
```

```{r}
human_genes <- readRDS(file = "./data/human_genes.rds")
```

We filter the previously loaded `human_genes` for only the circadian and driver
genes, then display those circadian genes in an interactive datatable.

```{r}
human_genes %>%
  dplyr::filter(gene_name %in% circadian_gene_list) %>%
  dplyr::arrange(gene_name) %>%
  as.data.frame() -> circadian_genes

# human_genes %>%
#   dplyr::filter(gene_name %in% driver_gene_list) %>%
#   dplyr::arrange(gene_name) %>%
#   as.data.frame() -> driver_genes

DT::datatable(circadian_genes)
```

## Depmap

We create an `ExperimentHub` object to pull relevant DepMap datasets:

    crispr: CRISPR-Cas9 essentiality (dependency) scores
    copyNumber: copy number alterations per gene/cell line
    TPM: transcript expression levels (RNA-seq)
    mutationCalls: mutation data
    metadata: cell-line metadata (including disease subtype).

```{r message=FALSE}
## create ExperimentHub query object
eh <- ExperimentHub()
query(eh, "depmap")
metadata <- eh[["EH7558"]]
crispr <- eh[["EH7554"]]
copyNumber <- eh[["EH7555"]]
TPM <- eh[["EH7556"]]
mutationCalls <- eh[["EH7557"]]
```

## Metadata

We subset DepMap metadata for melanoma lines using keywords ("melanoma") across
relevant columns. We display the resulting set of melanoma cell lines in a
datatable for easy scanning.

```{r error=FALSE, warning=FALSE}
metadata %>%
  dplyr::filter(grepl("melanoma", lineage_subtype) |
                grepl("elanoma", Cellosaurus_NCIt_disease) |
                grepl("elanoma", subtype_disease) | 
                grepl("elanoma", cell_line)) %>%
  dplyr::select(-contains("issues")) %>%
  as.data.frame() -> melanoma_metadata

DT::datatable(melanoma_metadata)
```

## CRISPR Dependency

We visualize how dependent these cell lines are on each circadian gene using [plotly](https://plotly.com/r/.
The dashed lines represent:

    Red line: the mean dependency across all circadian genes in just melanoma lines.
    Green line: the mean dependency across all lines and all genes in DepMap (global average).

```{r warning=FALSE, message=FALSE}
crispr %>%
  dplyr::filter(cell_line %in% melanoma_metadata$cell_line,
                gene_name %in% circadian_genes$gene_name,
                !is.na(cell_line)) %>%
  as.data.frame() -> crispr_df

crispr %>% 
  dplyr::filter(!is.na(dependency)) %>%
  as.data.frame() -> crispr_global_dependency
  
crispr_df %>%
  dplyr::filter(!is.na(dependency)) %>%
  dplyr::group_by(gene_name) %>%
  dplyr::summarize(mean_dependency = mean(dependency, na.remove = TRUE)) %>% 
  as.data.frame() -> crispr_gene_mean_dep

crispr_df %>%
  dplyr::left_join(crispr_gene_mean_dep, by = "gene_name") %>%
  dplyr::arrange(desc(mean_dependency)) %>%
  dplyr::mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  dplyr::select(-c(gene, cell_line)) %>% 
  dplyr::left_join(metadata, by = "depmap_id") %>%
  as.data.frame() -> crispr_gene_mean_dep_merged

crispr_gene_mean_dep_merged %>%
  ggplot(aes(x = gene_name, y = dependency, color = subtype_disease)) +
  geom_point(size = 0.75) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  geom_hline(yintercept = mean(crispr_gene_mean_dep$mean_dependency, na.remove = TRUE),
             color = "darkred", linetype = "dashed") +
  geom_hline(yintercept = mean(crispr_global_dependency$dependency, na.remove = TRUE),
             color = "darkgreen", linetype = "dashed") +
  xlab("circadian rhythm-related genes") +
  ggtitle("Circadian rhythm-related genes ranked by descending mean dependency score") -> p1
ggplotly(p1)
```

<!-- ```{r} -->
<!-- mutationCalls %>% -->
<!--   dplyr::filter(depmap_id %in% melanoma_metadata$depmap_id, -->
<!--                 gene_name %in% driver_genes$gene_name) %>% -->
<!--   as.data.frame() -> driver_df -->
<!-- ``` -->

<!-- ```{r} -->
<!-- crispr_gene_mean_dep_merged %>% -->
<!--   dplyr::arrange(desc(dependency)) %>% -->
<!--   dplyr::mutate(cell_line = factor(cell_line, levels = unique(cell_line))) %>% -->
<!--   ggplot(aes(x = cell_line, y = dependency, color = subtype_disease)) + -->
<!--   geom_point(size = 0.75) + -->
<!--   theme_classic() + -->
<!--     theme(axis.text.x=element_blank(), -->
<!--         axis.ticks.x=element_blank()) + -->
<!--   ggtitle("yolo") -->
<!-- ``` -->

We reshape (`pivot_wider`) the data so that `rows = genes`, `columns = melanoma`
cell lines, and `values = CRISPR` dependency scores. Then we call `pheatmap` to
visualize how essential each gene is across multiple lines in a grid layout.

**Interpretation: we don't observe a strong pattern in dependency scores in circadian genes in melanoma cell lines**

```{r warning=FALSE}
crispr_df %>%
  dplyr::select(gene_name, cell_line, dependency) %>%
  tidyr::pivot_wider(names_from = cell_line,
                     values_from = dependency) %>%
  tibble::column_to_rownames(var = "gene_name") %>%
  t() %>% as.data.frame() %>%
  tibble::rownames_to_column(var = "cell_line") %>%
  dplyr::left_join(melanoma_metadata, by = "cell_line") %>%
  as.data.frame() -> crispr_ann_df

data.frame(status = crispr_ann_df$primary_or_metastasis,
           subtype = crispr_ann_df$subtype_disease,
           site = crispr_ann_df$sample_collection_site,
           row.names = crispr_ann_df$cell_line) -> sample_col

crispr_ann_df %>%
  dplyr::select(cell_line:RORB) %>%
  tibble::column_to_rownames(var = "cell_line") %>%
  t() %>%
  pheatmap::pheatmap(
    annotation_col = sample_col,
    show_colnames = FALSE,
    border_color = NA,
    fontsize = 7,
    main = paste0("Heatmap of dependency scores for circadian rhythm-related ",
                  "genes in melanoma cell lines"))
```

```{r echo=FALSE}
rm(crispr)
# rm(crispr_df)
rm(crispr_global_dependency)
rm(crispr_gene_mean_dep)
rm(crispr_gene_mean_dep_merged)
```

### Copy Number

We check log2 copy-number values per gene. A dashed line at `y=1` (or `y=0`,
depending on your scale) might indicate a diploid reference. Higher or lower
values suggest possible amplification or deletion for these circadian genes in
melanoma cells. The genes are arranged by descending average copy number. For
more information how Depmap CNV is calculated, please refer to the [DepMap documentation](https://depmap.sanger.ac.uk/documentation/datasets/copy-number/).

**Interpretation: RORC average copy number is highest in melanoma, but also displays the greatest variation. The other circadian-related genes display variation, but their mean CNV is close to 1**

```{r}
copyNumber %>%
  dplyr::filter(cell_line %in% melanoma_metadata$cell_line,
                gene_name %in% circadian_genes$gene_name,
                !is.na(cell_line)) %>%
  as.data.frame() -> copy_number_df

copy_number_df %>%
  dplyr::arrange(desc(log_copy_number)) %>% 
  dplyr::mutate(gene_name = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = gene_name, y = log_copy_number, fill = gene_name)) +
    geom_violin() +
    geom_boxplot(width = 0.25) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          legend.position = "none") +
    geom_hline(yintercept = 1, color = "black", linetype = "dashed") +
    ggtitle(paste0("Log10 copy-number for circadian rhythm-related genes in",
                   " melanoma cell lines"))
```

### Gene Expression

We pivot from long to wide, building a matrix of TPM expression with genes as
rows and cell lines as columns. Then we optionally annotate columns with
metadata (`metastatic status`, `disease subtype`) and generate a `pheatmap` to
see expression clustering.

**Interpretation: Expression of PER2, RORA, RORB and RORC appears to be very low in all melanome cell lines. NPAS expression is the most elevated, while the remaining genes display expression somewhere in between.**

```{r fig.height=5, fig.width=9}
TPM %>%
  dplyr::filter(cell_line %in% melanoma_metadata$cell_line,
                gene_name %in% circadian_genes$gene_name,
                !is.na(cell_line)) %>%
  as.data.frame() -> tpm_df

tpm_df %>%
  dplyr::select(gene_name, cell_line, rna_expression) %>%
  tidyr::pivot_wider(names_from = cell_line,
                     values_from = rna_expression) %>%
  tibble::column_to_rownames(var = "gene_name") %>%
  t() %>% as.data.frame() %>%
  tibble::rownames_to_column(var = "cell_line") %>%
  dplyr::left_join(melanoma_metadata, by = "cell_line") %>%
  as.data.frame() -> tpm_ann_df

data.frame(status = tpm_ann_df$primary_or_metastasis,
           subtype = tpm_ann_df$subtype_disease,
           site = tpm_ann_df$sample_collection_site,
           row.names = tpm_ann_df$cell_line) -> sample_col

tpm_ann_df %>%
  dplyr::select(cell_line:RORB) %>%
  tibble::column_to_rownames(var = "cell_line") %>%
  t() %>%
  pheatmap::pheatmap(
    annotation_col = sample_col,
    fontsize = 7,
    border_color = NA,
    show_colnames = FALSE,
    main = paste0("Heatmap of log10 gene expression for circadian rhythm-related ",
                  "genes in melanoma cell lines"))
```

We join CRISPR dependency data with expression (`TPM`) for the same gene/cell line
pairs, then plot expression vs. dependency. The `stat_cor(method = "spearman")`
call adds a correlation value to see if higher expression correlates with lower
(or higher) essentiality.

**Interpretation: We don't observe a signfificant direct correlation between any circadian rhythm gene expression and CRISPR dependency scores.**

*Note: Normally, we would expect a relationship where increased gene expression displays increased dependency, if such a gene dependency exists.*

```{r fig.height=6}
tpm_df %>%
  dplyr::select(-c(gene, cell_line, entrez_id)) %>%
  dplyr::left_join(crispr_df, by = c("depmap_id", "gene_name")) %>%
  dplyr::filter(!is.na(cell_line)) %>%
  ggplot(aes(x = dependency, y = rna_expression, colour = gene_name)) +
    geom_point() +
    stat_smooth(method = "lm", se = TRUE, formula = y ~ poly(x, 1, raw = TRUE),
                color = "black", linetype = "dashed", fill = "gray") +
    ggpubr::stat_cor(method = "spearman", label.x = -0.5, label.y = 6) +
    ggtitle(paste0("Correlation between circadian-rhythm gene expression and ",
                   "CRISPR\ndependency scores in Depmap melanoma cell lines")) +
    facet_wrap(~gene_name, ncol = 3)
```

```{r echo=FALSE}
rm(TPM)
rm(tpm_df)
```

### Mutation Calls

We look at mutationCalls table, combine with metadata, and keep only those
entries with a circadian gene in melanoma cell lines.

```{r}
mutationCalls %>%
  dplyr::left_join(melanoma_metadata, by = "depmap_id") %>%
  dplyr::filter(cell_line %in% melanoma_metadata$cell_line,
                gene_name %in% circadian_genes$gene_name,
                !is.na(cell_line)) %>%
  as.data.frame() -> mutation_calls_df
```

We create a balloon plot to see how mutation classification (e.g.,
"Missense", "Nonsense", "Splice site") differs between primary vs. metastatic
cell lines.

<!-- ```{r} -->
<!-- mutation_calls_df %>% -->
<!--   ggplot(aes(x = var_class, fill = primary_or_metastasis)) + -->
<!--     geom_bar(position = "dodge") + -->
<!--     labs(x = "Mutation Type", y = "Number of Cell Lines") + -->
<!--     theme_classic() + -->
<!--     theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->

<!-- The following balloon plot displays the same information: -->

```{r}
table(mutation_calls_df$var_class, mutation_calls_df$primary_or_metastasis) %>%
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap elanoma circadian gene mutation type by metastatic status")
```

We create a balloon plot to see how mutation classification (e.g.,
"Missense", "Nonsense", "Splice site") differs between primary vs. metastatic cell lines.

<!-- ```{r} -->
<!-- mutation_calls_df %>% -->
<!--   ggplot(aes(x = var_class, fill = gene_name)) + -->
<!--     geom_bar(position = "stack") + -->
<!--     labs(x = "Mutation Type", y = "Number of Cell Lines") + -->
<!--     theme_classic() + -->
<!--     theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->

<!-- The following balloon plot displays the same information: -->

```{r}
table(mutation_calls_df$var_class, mutation_calls_df$gene_name) %>%
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap melanoma circadian gene mutation type by gene")
```

A balloon plot that partitions each bar by the circadian gene mutated, so
you can see which genes tend to have which var_class.

<!-- ```{r} -->
<!-- mutation_calls_df %>% -->
<!--   ggplot(aes(x = var_annotation, fill = primary_or_metastasis)) + -->
<!--     geom_bar(position = "dodge") + -->
<!--     labs(x = "Mutation Type", y = "Number of Cell Lines") + -->
<!--     theme_classic() + -->
<!--     theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->

```{r}
table(mutation_calls_df$var_annotation, mutation_calls_df$primary_or_metastasis) %>%
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap melanoma circadian gene mutation annotation by metastatic status")
```

We plot the variant annotation to see how it splits between primary vs metastatic lines.

<!-- ```{r} -->
<!-- mutation_calls_df %>% -->
<!--   ggplot(aes(x = var_annotation, fill = gene_name)) + -->
<!--     geom_bar(position = "stack") + -->
<!--     labs(x = "Mutation Type", y = "Number of Cell Lines") + -->
<!--     theme_classic() + -->
<!--     theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->

```{r}
table(mutation_calls_df$var_annotation, mutation_calls_df$gene_name) %>%
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap melanoma circadian gene mutation annotations by gene")
```

We create a custom column mutation_type that concatenates ref_allele and
alt_allele (e.g., C_to_T) to highlight known mutational signatures. The bar
chart categorizes them by primary vs. metastatic. This often reveals a high
frequency of UV-associated C>T transitions in melanoma.

*UV Exposure: In primary cutaneous melanomas, C to T (and symmetrically G to A) transitions at dipyrimidine sites are the prototypical "UV signature." Given the strong etiological of melanoma link with UV radiation, it is no surprise these transitions dominate.*

<!-- ```{r} -->
<!-- mutation_calls_df %>% -->
<!--   dplyr::mutate(mutation_type = paste0(ref_allele, "_to_", alt_allele)) %>% -->
<!--   ggplot(aes(x = mutation_type, fill = primary_or_metastasis)) + -->
<!--     geom_bar(position = "dodge") + -->
<!--     labs(x = "Mutation Type", y = "Number of Cell Lines") + -->
<!--     theme_classic() + -->
<!--     theme(axis.text.x = element_text(angle = 45, hjust = 1)) -->
<!-- ``` -->

```{r}
mutation_calls_df %>%
  dplyr::mutate(mutation_type = paste0(ref_allele, "_to_", alt_allele)) %>%
  as.data.frame() -> mutation_calls_df2

table(mutation_calls_df2$mutation_type, mutation_calls_df2$primary_or_metastasis) %>% 
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap melanoma circadian gene mutation transitions by metastatic status")
```
```{r}
mutation_calls_df %>%
  dplyr::mutate(mutation_type = paste0(ref_allele, "_to_", alt_allele)) %>%
  as.data.frame() -> mutation_calls_df2

table(mutation_calls_df2$mutation_type, mutation_calls_df2$gene_name) %>% 
  as.data.frame() %>% 
  ggballoonplot(fill = "value") +
    scale_fill_gradientn(colors = my_cols) +
    ggtitle("Depmap melanoma circadian gene mutation transitions by gene")
```

```{r echo=FALSE}
rm(mutationCalls)
rm(mutation_calls_df)
```

## Summary

1) Data Filtering: We demonstrated how to access and filter DepMap datasets to melanoma lines and circadian genes.
1) Visualizations: We demonstrated how to visualize Depmap data using bar plots, heatmaps, scatter plots, and interactive plots via plotly show how circadian gene dependency, copy number, expression, and mutations vary across these cell lines.
1) Future Directions: 
  + Integrate driver gene statuses (BRAF, NRAS, NF1, etc.) to see if certain circadian genes are more essential in specific mutational subgroups.
  + Integrate additional circadian-rhythm-related genes by compiling a list of genes from the following GO terms from Biomart: [GO:0007623 circadian rhythm](https://www.ebi.ac.uk/QuickGO/term/GO:0007623), [GO:0032922 circadian regulation of gene expression](https://www.ebi.ac.uk/QuickGO/term/GO:0032922), [GO:0042752 regulation of circadian rhythm](https://www.ebi.ac.uk/QuickGO/term/GO:0042752), [GO:0042753 positive regulation of circadian rhythm](https://www.ebi.ac.uk/QuickGO/term/GO:0042753) and [GO:0042754 negative regulation of circadian rhythm](https://www.ebi.ac.uk/QuickGO/term/GO:0042754).

This exploration underscores the utility of DepMap data and R-based visualization
workflows to generate hypotheses about circadian gene roles in melanoma cell-line
survival, copy number changes, and potentially UV-signature–associated mutations.

```{r}
sessionInfo()
```